File: src/lib/api/templates.ts
import { invoke, isTauri } from "@tauri-apps/api/core";
import type {
  TemplateCategory,
  TemplateDetail,
  TemplateFilter,
  TemplateInput,
  TemplateMetadata,
  TemplateVersionDetail,
  TemplateVersionMetadata,
} from "../types/templates";

export interface LocalTemplateImportResult {
  categoriesCreated: number;
  categoriesReused: number;
  templatesCreated: number;
  templatesSkipped: number;
  historiesSkipped: boolean;
}

export function importLocalTemplatesToCloud(): Promise<LocalTemplateImportResult> {
  return invoke<LocalTemplateImportResult>("cloud_import_local_templates");
}

export function listTemplateCategories(): Promise<TemplateCategory[]> {
  return isTauri()
    ? invoke<TemplateCategory[]>("cloud_list_template_categories").then((items) => items.map(normalizeCategory))
    : Promise.resolve([]);
}

export function createTemplateCategory(name: string, parentId?: number): Promise<TemplateCategory> {
  return invoke<TemplateCategory>("cloud_create_template_category", { name, parentId: parentId ?? null })
    .then(normalizeCategory);
}

export function renameTemplateCategory(id: number, name: string): Promise<void> {
  return invoke<void>("cloud_rename_template_category", { id, name });
}

export function deleteTemplateCategory(id: number): Promise<void> {
  return invoke<void>("cloud_delete_template_category", { id });
}

export function moveTemplateCategory(
  id: number,
  parentId: number | undefined,
  targetIndex: number,
): Promise<void> {
  return invoke<void>("cloud_move_template_category", { id, parentId: parentId ?? null, targetIndex });
}

export function listTemplates(filter: TemplateFilter): Promise<TemplateMetadata[]> {
  return isTauri()
    ? invoke<TemplateMetadata[]>("cloud_list_templates", { filter }).then((items) => items.map(normalizeMetadata))
    : Promise.resolve([]);
}

export function searchTemplateCompletions(
  query: string,
  limit = 20,
): Promise<TemplateDetail[]> {
  return isTauri()
    ? invoke<TemplateDetail[]>("cloud_search_template_completions", { query, limit })
      .then((items) => items.map(normalizeDetail))
    : Promise.resolve([]);
}

export function getTemplate(id: number): Promise<TemplateDetail> {
  return invoke<TemplateDetail>("cloud_get_template", { id }).then(normalizeDetail);
}

export function getTemplates(ids: number[]): Promise<TemplateDetail[]> {
  return isTauri()
    ? invoke<TemplateDetail[]>("cloud_get_templates", { ids }).then((items) => items.map(normalizeDetail))
    : Promise.resolve([]);
}

export function createTemplate(input: TemplateInput): Promise<TemplateDetail> {
  return invoke<TemplateDetail>("cloud_create_template", { input }).then(normalizeDetail);
}

export function updateTemplate(id: number, input: TemplateInput): Promise<TemplateDetail> {
  return invoke<TemplateDetail>("cloud_update_template", { id, input }).then(normalizeDetail);
}

export function deleteTemplate(id: number): Promise<void> {
  return invoke<void>("cloud_delete_template", { id });
}

export function setTemplateFavorite(id: number, favorite: boolean): Promise<void> {
  return invoke<void>("cloud_set_template_favorite", { id, favorite });
}

export function recordTemplateUse(id: number): Promise<void> {
  return invoke<void>("cloud_record_template_use", { id });
}

export function moveTemplate(
  id: number,
  categoryId: number | undefined,
  targetIndex: number,
): Promise<void> {
  return invoke<void>("cloud_move_template", { id, categoryId: categoryId ?? null, targetIndex });
}

export function listTemplateVersions(templateId: number): Promise<TemplateVersionMetadata[]> {
  void templateId;
  return Promise.resolve([]);
}

export function getTemplateVersion(versionId: number): Promise<TemplateVersionDetail> {
  void versionId;
  return Promise.reject(new Error("云模板第一版暂不提供历史版本。"));
}

export function deleteTemplateVersion(templateId: number, versionId: number): Promise<void> {
  void templateId;
  void versionId;
  return Promise.reject(new Error("云模板第一版暂不提供历史版本。"));
}

export function restoreTemplateVersion(
  templateId: number,
  versionId: number,
): Promise<TemplateDetail> {
  void templateId;
  void versionId;
  return Promise.reject(new Error("云模板第一版暂不提供历史版本。"));
}

function normalizeCategory(category: TemplateCategory): TemplateCategory {
  return { ...category, parentId: category.parentId ?? undefined };
}

function normalizeMetadata(template: TemplateMetadata): TemplateMetadata {
  return {
    ...template,
    categoryId: template.categoryId ?? undefined,
    lastUsed: template.lastUsed ?? undefined,
  };
}

function normalizeDetail(template: TemplateDetail): TemplateDetail {
  return { ...normalizeMetadata(template), code: template.code };
}

File: src-tauri/src/commands/templates.rs
use tauri::State;

use crate::{
    error::CommandError,
    state::AppState,
    templates::{
        self as template_core, TemplateCategory, TemplateDetail, TemplateFilter, TemplateInput,
        TemplateMetadata, TemplateVersionDetail, TemplateVersionMetadata,
    },
};

#[tauri::command(async)]
pub fn list_template_categories(
    state: State<'_, AppState>,
) -> Result<Vec<TemplateCategory>, CommandError> {
    template_core::list_categories(&state.paths.database_file).map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn create_template_category(
    name: String,
    parent_id: Option<i64>,
    state: State<'_, AppState>,
) -> Result<TemplateCategory, CommandError> {
    template_core::create_category(&state.paths.database_file, &name, parent_id)
        .map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn rename_template_category(
    id: i64,
    name: String,
    state: State<'_, AppState>,
) -> Result<(), CommandError> {
    template_core::rename_category(&state.paths.database_file, id, &name)
        .map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn delete_template_category(id: i64, state: State<'_, AppState>) -> Result<(), CommandError> {
    template_core::delete_category(&state.paths.database_file, id).map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn move_template_category(
    id: i64,
    parent_id: Option<i64>,
    target_index: usize,
    state: State<'_, AppState>,
) -> Result<(), CommandError> {
    template_core::move_category(&state.paths.database_file, id, parent_id, target_index)
        .map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn list_templates(
    filter: TemplateFilter,
    state: State<'_, AppState>,
) -> Result<Vec<TemplateMetadata>, CommandError> {
    template_core::list_templates(&state.paths.database_file, &filter).map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn get_template(id: i64, state: State<'_, AppState>) -> Result<TemplateDetail, CommandError> {
    template_core::get_template(&state.paths.database_file, id).map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn create_template(
    input: TemplateInput,
    state: State<'_, AppState>,
) -> Result<TemplateDetail, CommandError> {
    template_core::create_template(&state.paths.database_file, &input).map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn update_template(
    id: i64,
    input: TemplateInput,
    state: State<'_, AppState>,
) -> Result<TemplateDetail, CommandError> {
    template_core::update_template(&state.paths.database_file, id, &input)
        .map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn delete_template(id: i64, state: State<'_, AppState>) -> Result<(), CommandError> {
    template_core::delete_template(&state.paths.database_file, id).map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn set_template_favorite(
    id: i64,
    favorite: bool,
    state: State<'_, AppState>,
) -> Result<(), CommandError> {
    template_core::set_favorite(&state.paths.database_file, id, favorite)
        .map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn record_template_use(id: i64, state: State<'_, AppState>) -> Result<(), CommandError> {
    template_core::record_use(&state.paths.database_file, id).map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn move_template(
    id: i64,
    category_id: Option<i64>,
    target_index: usize,
    state: State<'_, AppState>,
) -> Result<(), CommandError> {
    template_core::move_template(&state.paths.database_file, id, category_id, target_index)
        .map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn list_template_versions(
    template_id: i64,
    state: State<'_, AppState>,
) -> Result<Vec<TemplateVersionMetadata>, CommandError> {
    template_core::list_versions(&state.paths.database_file, template_id)
        .map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn get_template_version(
    version_id: i64,
    state: State<'_, AppState>,
) -> Result<TemplateVersionDetail, CommandError> {
    template_core::get_version(&state.paths.database_file, version_id).map_err(CommandError::from)
}

#[tauri::command(async)]
pub fn restore_template_version(
    template_id: i64,
    version_id: i64,
    state: State<'_, AppState>,
) -> Result<TemplateDetail, CommandError> {
    template_core::restore_version(&state.paths.database_file, template_id, version_id)
        .map_err(CommandError::from)
}
